net/wgpeerselector/files/usr/bin/wgpeerselector

```diff
@@ -0,0 +1,501 @@
+	log(syslog.LOG_INFO, "Removing peer " .. self.name .. " from " .. self.iface ..".")
+	if os.execute(command) ~= 0 then
+		log(syslog.LOG_WARNING, "Couldn't remove peer " .. self.name .. " from " .. self.iface .. ".")
+		return false
+	end
+	for _, allowed_ip in pairs(self.allowed_ips) do
+		if os.execute("ip route del '" .. allowed_ip .. "' dev '" .. self.iface .. "'") ~= 0 then
+			log(syslog.LOG_WARNING, "Couldn't del route " .. allowed_ip .. " from " .. self.iface .. ".")
+		end
+	end
+	self.established_at = 0
+	return true
+end
+
+function WGPeer:update_stats_from_kernel()
+	local res = execute('wg show "' .. self.iface .. '" dump')
+
+	for line in res:gmatch("([^\n]*)[\n]?") do
+		local public_key = string.match(line, "^([^\t]+)\t");
+		-- skip potential other peers
+		if public_key == self.public_key then
+			-- skip the other parts
+			self.latest_handshake, self.tx_bytes, self.rx_bytes =
+				string.match(line, "^[^\t]+\t[^\t]+\t[^\t]+\t[^\t]+\t(%d+)\t(%d+)\t(%d+)\t[^\t]+$")
+
+			if self.established_at == 0 then
+				self.established_at = self.latest_handshake
+			end
+
+			return true
+		end
+	end
+
+	return false
+end
+
+function WGPeer:established_time()
+	if not self:has_recent_handshake() then
+		return nil
+	end
+	return (time.time() - self.established_at)
+end
+
+function WGPeer:has_recent_handshake()
+	-- wireguard handshakes are sent at least every 2 minutes, if there is
+	-- payload traffic.
+	return (time.time() - self.latest_handshake) < 150
+end
+
+local WGPeerSelector = {}
+
+function WGPeerSelector:new(o)
+	o = o or {}   -- create object if user does not provide one
+	setmetatable(o, self)
+	self.__index = self
+	-- some defaults
+	self.queued_peers = {}
+	self.peers = {}
+	return o
+end
+
+function WGPeerSelector:load_peers_from_uci()
+	uci:foreach('wgpeerselector', 'peer', function(peer)
+		if uci:get_bool('wgpeerselector', peer['.name'], 'enabled')
+		   and peer.ifname == self.iface and peer.public_key then
+			-- TODO: error on wrong schema?
+			table.insert(self.peers, WGPeer:new{
+				name = peer['.name'],
+				iface = self.iface,
+				public_key = peer.public_key,
+				endpoint = peer.endpoint,
+				allowed_ips = peer.allowed_ips or {},
+				persistent_keepalive = peer.persistent_keepalive or 0,
+			})
+		end
+	end)
+
+	if #self.peers < 1 then
+		log(syslog.LOG_ERR, "No peers defined. Exiting!", true)
+		os.exit(1)
+	end
+
+	log(syslog.LOG_INFO, 'Loaded '..tostring(#self.peers)..' peers.')
+end
+
+function WGPeerSelector:queue_all_peers()
+	-- copy table
+	for _, peer in pairs(self.peers) do
+		table.insert(self.queued_peers, peer)
+	end
+end
+
+function WGPeerSelector:queue_pop_random_peer()
+	return table.remove(self.queued_peers, math.random(#self.queued_peers))
+end
+
+function WGPeerSelector:try_connect_to_peer(peer, timeout)
+	if not peer:install_to_kernel() then
+		return false
+	end
+
+	sleep(timeout)
+	peer:update_stats_from_kernel()
+
+	local connection_successful = peer:has_recent_handshake()
```

Comment:
Checking for a recent handshake immediately after `wg set` implies that you have persistent keepalive enabled or something else generating traffic, right?

Reply:
Yes, for Gluon we have the mesh protocol that continuously generates traffic every X seconds. Also, I added a warning about this potential issue in the README ([here](https://github.com/freifunk-gluon/packages/pull/240/files/3a29b1fa0177ffcdecd3f15a3f51b448a71cb011#diff-e8d18fa5462091b44ec1c5a859daf30b8b1a745b6f4e5c742932cf22d529ca19R93-R97)).